File: golang/utils/compare_string.go
package utils

func CompareString(a, b string) bool {
	if len(a) != len(b) {
		return false
	}

	for i := 0; i < len(a); i++ {
		if a[i] != b[i] {
			return false
		}
	}

	return true
}

func SafeCompareString(a, b string) bool {
	if len(a) != len(b) {
		return false
	}
	lenA := len(a)
	var result byte = 0
	for i := 0; i < lenA; i++ {
		result |= a[i] ^ b[i]
	}

	return result == 0
}

File: typescript/src/utils/compare-string.ts
export function compareString(a: unknown, b: unknown): boolean {
    if (!a || !b || typeof a !== "string" || typeof b !== "string") {
        return false;
    }

    const lenA = a.length;
    if (lenA !== b.length) {
        return false;
    }

    for (let i = 0; i < lenA; i++) {
        if (!(a.charCodeAt(i) === b.charCodeAt(i))) {
            return false;
        }
    }

    return true;
}

export function safeCompareString(a: unknown, b: unknown): boolean {
    if (!a || !b || typeof a !== "string" || typeof b !== "string") {
        return false;
    }

    const lenA = a.length;
    if (lenA !== b.length) {
        return false;
    }

    let result: number = 0;

    for (let i = 0; i < lenA; i++) {
        result |= a.charCodeAt(i) ^ b.charCodeAt(i);
    }

    return result === 0;
}
